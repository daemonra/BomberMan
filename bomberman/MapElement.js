import SpecialAbillity from "./SpecialAbillity.js";
import Invincible from "./Invincible.js";
import Obstacle from "./Obstacle.js";
import Rollerskate from "./Rollerskate.js";
import IncreaseBombsNumber from "./IncreaseBombsNumber.js";
import IncreaseBlastRange from "./IncreaseBlastRange.js"
import Ghost from "./Ghost.js";
import Detonator from "./Detonator.js";
import DecreaseBlastRange from "./DecreaseBlastRange.js";
import DecreaseSpeed from "./DecreaseSpeed.js";
import PlaceBombsASAP from "./PlaceBombsASAP.js";
import CantPlaceBombs from "./CantPlaceBombs.js";

export default class MapElement {
    
    static Field = new MapElement("field");
    static Wall = new MapElement("wall");
    static Monster = new MapElement("monster");
    static Player1 = new MapElement("player playerOne", null, null, []);
    static Player2 = new MapElement("player playerTwo", null, null, []);
    static Bomb = new MapElement("bomb");
    static BombInPlace = new MapElement("bombip")
    static Explosion = new MapElement("explosion");
    static Box = new MapElement("box");
    static BoxEmpty = new MapElement("box")
    static BoxInPlace = new MapElement("boxip")

    //curses
    static Curse = new MapElement("curse", null, [DecreaseBlastRange.instance(), DecreaseSpeed.instance(), PlaceBombsASAP.instance(), CantPlaceBombs.instance()])
    
    //powerups
    static Invincible = new MapElement("powerup invincible", Invincible.instance());
    static Detonator = new MapElement("powerup detonator", Detonator.instance());
    static Rollerskate = new MapElement("powerup rollerskate", Rollerskate.instance());
    static Ghost = new MapElement("powerup ghost", Ghost.instance());
    static PlaceBoxes = new MapElement("powerup placeBoxes", Obstacle.instance());
    static IncreaseBlastRange = new MapElement("powerup increaseBlastRange", IncreaseBlastRange.instance());
    static IncreaseBombsNumber = new MapElement("powerup increaseBombsNumber", IncreaseBombsNumber.instance());

    constructor(value, instance = null, cursesList = null, extraClasses = null) {  
        this.classType = value;
        this.instance = instance;
        this.cursesList = cursesList;
        this.extraClasses = extraClasses;
    }

    static allPowerups(){
        let arr=[]
        arr.push(this.Invincible)
        arr.push(this.Detonator)
        arr.push(this.Rollerskate)
        arr.push(this.Ghost)
        arr.push(this.PlaceBoxes)
        arr.push(this.IncreaseBlastRange)
        arr.push(this.IncreaseBombsNumber)

        return arr;
    }

    static randomizePowerups() {
        return this.allPowerups()[Math.floor(Math.random() * (6+1))];
    }
}