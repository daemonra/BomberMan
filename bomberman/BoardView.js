import MapElement from './MapElement.js';
import Invincible from './Invincible.js';

export default class BoardView {
    
    constructor() {}

    showModule(moduleId) {
        const module = document.getElementById(moduleId);
        if (module) {
            module.classList.remove('hidden'); 
        }
    }
    
    hideModule(moduleId) {
        const module = document.getElementById(moduleId);
        if (module) {
            module.classList.add('hidden'); 
        }
    }

    renderMatrix(matrix) {
        let str = ""
    
        for (let i = 0 ; i < matrix.length; i++) {
            
            str+="<tr>"
    
            for (let j = 0 ; j < matrix[i].length; j++) {
                let extraClasses = "" 
                if (matrix[i][j].extraClasses !== null) {
                    extraClasses = matrix[i][j].extraClasses.join(" ")
                }
                str+="<td class='grid "+matrix[i][j].classType+" "+extraClasses+"' data-i='"+i+"' data-j='"+j+"'><span></span></td>"
            }
    
            str+="</tr>"
        }
        document.querySelector('table#map').innerHTML = str;
    }

    formatedTime(seconds) {
        const minutes = Math.floor(seconds / 60);
        const remainingSeconds = seconds % 60;
        
        const formattedMinutes = minutes.toString().padStart(1, '0'); // Ensure at least one digit
        const formattedSeconds = remainingSeconds.toString().padStart(2, '0');
        
        return `${formattedMinutes}:${formattedSeconds}`;
    }

    updateGameTimer(s) {
        document.getElementById("gameTimer").innerHTML = this.formatedTime(s);
    }

    updateBattleRoyaleTime(s) {
        document.getElementById("BattleRoyaleTimer").innerHTML = this.formatedTime(s);
    }

    displayExitPopUp() {} // continue like this 
}