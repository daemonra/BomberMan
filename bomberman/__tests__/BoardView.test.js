/**
 * @jest-environment jsdom
 */

import BoardView from '../BoardView.js';
import MapElement from '../MapElement.js';

describe('BoardView', () => {
  let boardView;

  beforeEach(() => {
    // Create a new instance of BoardView before each test
    boardView = new BoardView();
  });

  test('showModule should display the module with the given ID', () => {
    // Mock the necessary elements in the DOM
    document.body.innerHTML = '<div id="module1" style="display: flex;" class="hidden"></div>';
    const moduleId = 'module1';
    const moduleElement = document.getElementById(moduleId);     

    // Call the showModule method
    boardView.showModule(moduleId);

    // Expect the module to be displayed
    expect(moduleElement.style.display).toBe('flex');
  });


  test('renderMatrix should update the table#map with the provided matrix', () => {
    // Mock the necessary elements in the DOM
    document.body.innerHTML = '<table id="map"></table>';
    const matrix = [
      [MapElement.Field, MapElement.Wall],
      [MapElement.Player1, MapElement.Player2]
    ];

    // Call the renderMatrix method
    boardView.renderMatrix(matrix);

    // Expect the table#map to be updated with the matrix
    const table = document.querySelector('table#map');
    expect(table.innerHTML).toMatchSnapshot();
  });

  test('displayExitPopUp should do something', () => {
    // Mock any necessary elements or DOM manipulation needed for this method
    
    // Call the displayExitPopUp method
    boardView.displayExitPopUp();

    // Assert expectations here
  });

  // Add tests here if necessary
});
