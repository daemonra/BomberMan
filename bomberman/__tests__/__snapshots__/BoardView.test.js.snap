// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardView renderMatrix should update the table#map with the provided matrix 1`] = `"<tbody><tr><td class="grid field " data-i="0" data-j="0"><span></span></td><td class="grid wall " data-i="0" data-j="1"><span></span></td></tr><tr><td class="grid player playerOne " data-i="1" data-j="0"><span></span></td><td class="grid player playerTwo " data-i="1" data-j="1"><span></span></td></tr></tbody>"`;
